### DISCUSSION TITLE
Suspicion That M5STACK ATOM RS-485 Base Does Not Stand Up to the Load

### DISCUSSION THREAD

**Discussion by Code394** [omerfaruk-aran/esphome_samsung_hvac_bus#197] [General]
Peace and blessings,

Two days ago, without any configuration changes, ESPHome updates, or physical adjustments to the cables, I have been unable to send commands to control the air conditioners. This issue occurs both through Home Assistant and the ESPHome web interface, though I am still able to receive data regarding the air conditioners' status without any problems.
I will note that very rarely, I do manage to send a single command to the air conditioners.

I suspect that the RS-485 adapter may have been damaged due to the load (I have 21 indoor units and two outdoor units). In the past, when I used S-NET, my USB to RS-485 adapters would fail until I purchased [this](https://www.waveshare.com/product/iot-communication/wired-comm-converter/usb-to-rs232-uart-rs485/usb-to-rs232-485-ttl.htm) model, which has been reliable. Even now, I can send commands through S-NET without issues.

I tried connecting the M5STACK RS-485 directly to another RS-485 converter to verify that data transmission and reception were working correctly, and they did. However, I still suspect that the M5STACK struggles to handle the load when dealing with the heavy traffic from the air conditioning units.

I also tried several troubleshooting steps, including downgrading the ESPHome version, using the ESPHome dev version, resetting the board to default, restarting the air conditioners, and disconnecting and reconnecting the cables, but none of these helped.

Could you help me pinpoint the source of the problem?

(I am already considering buying an ESP32 board with connection Ethernet [like this one](https://he.aliexpress.com/item/1005006485369425.html) and connecting it to a robust TTL to RS-485 converter, like [this](https://www.waveshare.com/product/iot-communication/wired-comm-converter/usb-to-rs232-uart-rs485/multibus-converter.htm), to create a stable wired setup that will last for years.)

**Comment by brazoayeye** [comment]
it's very strange, the baud rate is very low (9600) and i found impossible that the hardware is not able to handle it. 

Have you tested with SNET (PC), an 232TTL and M5STACK RS-485 ? That could confirm that it's not an hardware related issue. 

Imho it could be something related to the software, i really don't know how espHome handle concurrency (and/or multithreading, reentrancy) and how high latency processes (wifi, web) could impact serial ones. Double check if the conf is the latest used and try to remove logs.

Haven't you ever noticed comm problems between samsung devices?

**Comment by Code394** [comment]
The communication between the air conditioners is working perfectly, and I am able to send commands without any issues using the S-NET system. However, I currently do not have another TTL to RS-485 adapter to test with. The only adapters I have available are a USB to TTL adapter and a USB to RS-485 adapter.

Is there any way I can use these adapters to help identify the source of the problem?

If the issue isn't related to network load, it may be caused by voltage on the RS-485 line disturbances affecting the M5STACK ATOM RS-485 converter, as I understand it lacks built-in protections against such disturbances.

Thank you for your assistance.

**Reply by brazoayeye** [reply]
Try to Connect PC -> USB to TTL -> M5STACK RS-485 and check with snet if it works. If you can make it works you are sure that the M5STACK RS-485 works properly and the problem is related to the ESP part.

Another test i would do is to print with esp only the raw hex recived data while recording the same with a PC and an USB to 485 device. You should see exactly the same raw data acquired if the harware is working properly. 

With the pc I usually use termite where you can enable HEX view  (putty can't)

Maybe you can check power supplies with an oscilloscope. Esp consumption is quite high while the wifi is operating, and I don't know how it's connected.
For the test you can connect a 3.3V power supply, or an USB 5V 1A power supply AND a 3.3V voltage regulator

It's up to you to decide if to start from software or from hardware side.

**Comment by Code394** [comment]
I was talking about voltage disturbances from the RS-485 line.
I will try the test you suggested later today and update.

**Comment by Code394** [comment]
I connected PC -> USB to TTL -> M5STACK RS-485. As expected, the S-NET could only read data and was unable to send commands. This confirms that the issue lies with the M5STACK RS-485 adapter.

I’ve decided that I need to purchase a high-quality TTL to RS-485 converter with protections on the RS-485 channel. While M5STACK offers [an isolated RS-485 converter](https://shop.m5stack.com/products/isolated-rs485-unit) with protection, I’d prefer to invest in [Waveshares ](https://www.waveshare.com/product/iot-communication/wired-comm-converter/usb-to-rs232-uart-rs485/multibus-converter.htm)high-end TTL to RS-485 converter. I already have another Waveshare converter that has lasted for years without issues. Even though some of Waveshare’s lower-end USB to RS-485 converters [like this one](https://www.waveshare.com/usb-to-rs485.htm?___SID=U) failed quickly, the high-end model has proven to be reliable.

Thanks again for your help.
